File: functions/pyaes/handler.py
from time import time
import random
import string
import pyaes
import json

# {
#     "length_of_message": 1000,
#     "num_of_iterations": 100
# }


def generate(length):
    letters = string.ascii_lowercase + string.digits
    return ''.join(random.choice(letters) for i in range(length))


def handle(event, context):
    start = time()
    req = json.loads(event.body.decode())
    length_of_message = req['length_of_message']
    num_of_iterations = req['num_of_iterations']

    message = generate(length_of_message)

    # 128-bit key (16 bytes)
    KEY = b'\xa1\xf6%\x8c\x87}_\xcd\x89dHE8\xbf\xc9,'

    for loops in range(num_of_iterations):
        aes = pyaes.AESModeOfOperationCTR(KEY)
        ciphertext = aes.encrypt(message)

        aes = pyaes.AESModeOfOperationCTR(KEY)
        plaintext = aes.decrypt(ciphertext)
        aes = None

    latency = time() - start

    return {
        "statusCode": 200,
        "body": {
            'latency': latency,
            'data': ''
        }
    }

File: functions/crypto/handler.js
"use strict";

const crypto = require("crypto");

function generate(length) {
  const characters = "abcdefghijklmnopqrstuvwxyz0123456789";
  let result = "";
  for (let i = 0; i < length; i++) {
    const randomIndex = Math.floor(Math.random() * characters.length);
    result += characters[randomIndex];
  }
  return result;
}

module.exports = async (event, context) => {
  const req = typeof event.body === "object" ? event.body : JSON.parse(event.body);
  const lengthOfMessage = req.length_of_message;
  const numOfIterations = req.num_of_iterations;

  const start = Date.now();
  const message = generate(lengthOfMessage);

  // 128-bit key (16 bytes)
  const KEY = Buffer.from([
    0xa1, 0xf6, 0x25, 0x8c, 0x87, 0x7d, 0x5f, 0xcd, 0x89, 0x64, 0x48, 0x45, 0x38, 0xbf, 0xc9, 0x2c,
  ]);

  let plaintext = "";
  for (let loops = 0; loops < numOfIterations; loops++) {
    const aes = crypto.createCipheriv("aes-128-ctr", KEY, Buffer.alloc(16, 0)); // Use an all-zero IV for CTR mode
    const ciphertext = aes.update(message, "utf8", "hex") + aes.final("hex");

    const decipher = crypto.createDecipheriv("aes-128-ctr", KEY, Buffer.alloc(16, 0));
    plaintext = decipher.update(ciphertext, "hex", "utf8") + decipher.final("utf8");
  }
  const latency = Date.now() - start;

  return context
    .status(200)
    .succeed({
      latency: latency / 1000,
      data: plaintext,
    });
};
